`yaml_of_string` (and `yaml_to_string`) fails with big inputs

The following program tries to parses and prints YAML data repeatedly, whose sizes gradually increase.  The program fails saying `doc_end failed` when the input size exceeds at some point:

```
(* ocamlfind ocamlc -package yaml x.ml -linkpkg *)

let () =
  let f i = 
    let buf = Buffer.create 1024 in
    Buffer.add_string buf "records:\n";
    for i = 1 to i do
      Buffer.add_string buf @@ Printf.sprintf "  - delay: %d\n" i;
      Buffer.add_string buf "    context: \"xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\"\n";
    done;
    let s = Buffer.contents buf in
    match Yaml.yaml_of_string s with
    | Error e -> Format.eprintf "%d: yaml_of_string: %a@." i Rresult.R.pp_msg e
    | Ok y -> 
        match Yaml.yaml_to_string y with
        | Error e -> Format.eprintf "%d: yaml_to_string: %a@." i Rresult.R.pp_msg e
        | Ok s -> Format.eprintf "%d: ok@." i
  in
  for i = 1 to 200 do
    f i
  done
```

Here is the output:

```
1: ok
2: ok
3: ok
4: ok
5: ok
6: ok
7: ok
8: ok
9: ok
10: ok
11: ok
12: ok
13: ok
14: ok
15: ok
16: ok
17: ok
18: ok
19: ok
20: ok
21: ok
22: ok
23: ok
24: ok
25: ok
26: ok
27: ok
28: ok
29: ok
30: ok
31: ok
32: ok
33: ok
34: ok
35: ok
36: ok
37: ok
38: ok
39: ok
40: ok
41: ok
42: ok
43: ok
44: ok
45: ok
46: ok
47: ok
48: ok
49: ok
50: ok
51: ok
52: ok
53: ok
54: ok
55: ok
56: ok
57: ok
58: ok
59: ok
60: ok
61: ok
62: ok
63: ok
64: ok
65: ok
66: ok
67: ok
68: ok
69: ok
70: ok
71: ok
72: ok
73: ok
74: ok
75: ok
76: ok
77: ok
78: ok
79: ok
80: ok
81: ok
82: ok
83: ok
84: ok
85: ok
86: ok
87: ok
88: ok
89: ok
90: ok
91: ok
92: ok
93: ok
94: ok
95: ok
96: ok
97: ok
98: ok
99: ok
100: ok
101: ok
102: ok
103: ok
104: ok
105: ok
106: ok
107: ok
108: ok
109: ok
110: ok
111: ok
112: ok
113: ok
114: ok
115: ok
116: ok
117: ok
118: ok
119: ok
120: ok
121: ok
122: ok
123: ok
124: ok
125: ok
126: ok
127: ok
128: ok
129: ok
130: ok
131: ok
132: ok
133: ok
134: ok
135: ok
136: ok
137: ok
138: ok
139: ok
140: ok
141: ok
142: ok
143: ok
144: ok
145: ok
146: ok
147: ok
148: ok
149: ok
150: ok
151: ok
152: ok
153: ok
154: ok
155: ok
156: ok
157: ok
158: ok
159: ok
160: ok
161: ok
162: ok
163: ok
164: ok
165: ok
166: ok
167: ok
168: ok
169: ok
170: ok
171: ok
172: ok
173: ok
174: ok
175: yaml_of_string: error calling parser
176: yaml_of_string: error calling parser
177: ok
178: yaml_of_string: error calling parser
179: yaml_of_string: error calling parser
180: ok
181: yaml_of_string: error calling parser
182: yaml_of_string: error calling parser
183: yaml_of_string: error calling parser
184: yaml_of_string: error calling parser
185: yaml_of_string: error calling parser
186: yaml_of_string: error calling parser
187: yaml_of_string: error calling parser
188: yaml_of_string: error calling parser
189: yaml_of_string: error calling parser
190: yaml_of_string: error calling parser
191: yaml_of_string: error calling parser
192: yaml_of_string: error calling parser
193: yaml_of_string: error calling parser
194: yaml_of_string: error calling parser
195: yaml_of_string: error calling parser
196: yaml_of_string: error calling parser
197: yaml_of_string: error calling parser
198: yaml_of_string: error calling parser
199: yaml_of_string: error calling parser
200: yaml_of_string: error calling parser
```

I tried this program with `yaml.0.2.1` on Mac OS X and Ubuntu and got the same results.

I also got `scalar failed` error when I used `yaml_to_string` against a large YAML data.  Maybe something is wrong around C FFI.

I wanted to follow up with some findings so far. So I'll include my approach and outcomes.

### Approach
To find a valid very large YAML file, I utilized the k8s swagger API and converted it from JSON to YAML with a simple python script. With the whole file I was receiving SEGV faults, to help shrink the problem I remove much of the YAML file, keeping only enough to still present with a failure. It ends up being ~350 lines, so much smaller then I originally expected.

Next, reviewed the code, to get an understanding of where the issues may be coming from. The most obvious place was in the CType bindings, and the memory management of the libyaml library which presents several `_delete` function, I don't believe are currently being used.

#### Attempts
First, I added in the `_delete` calls for events, parsers and emitters, While I believe this is a net positive for the library I wasn't able to observe any difference in the result or performance. I also adds some additional complexity managing the lifecycle of the parser and emitter, so unless the benefits can be measured, I'm unsure if it provides any value.

Second, I observed that depending on the order of testing the `too_large.yml` file, I would receive different results for the some of the existing tests. To help investigate this I added a shuffle to the files and would run the alcotests in a for loop. It seems when the "too_large.yml" is first, it causes additional test failures. I'll post the results shortly. I'm currently working under the assumption, that this supports the memory management of calls to libyaml hypothesis.

##### Test output
Note that test results are from a branch with many small changes to logging. Read the next to attempt for a description of why.
```
Done: 176/178 (jobs: 1)Testing Yaml.                                     
This run has ID `C4219E06-58B9-4758-963F-A55D463F4071`.               
[OK]                parse_event_test_set          0   yaml/bomb.yml.    
[ERROR]             parse_event_test_set          1   yaml/too_large.yml.                                                                                                                                                                                                                  
[OK]                parse_event_test_set          2   yaml/anchor.yml.   
[OK]                parse_event_test_set          3   yaml/cohttp.yml.
[OK]                parse_event_test_set          4   yaml/linuxkit.yml. 
[OK]                parse_event_test_set          5   yaml/seq.yml.   
[ERROR]             parse_of_string               0   yaml/too_large.yml. 
[OK]                parse_of_string               1   yaml/anchor.yml.  
[OK]                parse_of_string               2   yaml/cohttp.yml.
[OK]                parse_of_string               3   yaml/linuxkit.yml.
[OK]                parse_of_string               4   yaml/seq.yml.
[OK]                reflect                       0   yaml/bomb.yml.
[ERROR]             reflect                       1   yaml/too_large.yml.
[OK]                reflect                       2   yaml/anchor.yml.
[OK]                reflect                       3   yaml/cohttp.yml.
[OK]                reflect                       4   yaml/linuxkit.yml.
[OK]                reflect                       5   yaml/seq.yml.
[OK]                emit                          0   emit.
[OK]                version                       0   version.
[OK]                quoted_scalars                0   quoted bool.
[OK]                quoted_scalars                1   quoted null.
[OK]                quoted_scalars                2   quoted float.
[OK]                quoted_scalars                3   quoted string.
[OK]                quoted_scalars                4   quoted int.
[OK]                quoted_scalars                5   plain int.
[OK]                quoted_scalars                6   plain bool.
[OK]                quoted_scalars                7   plain string.
[OK]                quoted_scalars                8   plain string with quote.
[OK]                quoted_scalars                9   plain string with int and quote.
[OK]                quoted_scalars               10   double quoted string.
[OK]                quoted_scalars               11   partially double quoted string.
[OK]                yaml_equal                    0   two null equal.
[OK]                yaml_equal                    1   two floats equal.
[OK]                yaml_equal                    2   two floats not equal.
[OK]                yaml_equal                    3   two bools equal.
[OK]                yaml_equal                    4   two strings equal.
[OK]                yaml_equal                    5   two strings not equal.
[OK]                yaml_equal                    6   two arrays equal.
[OK]                yaml_equal                    7   two arrays not equal.
[OK]                yaml_equal                    8   two objects equal.
[OK]                yaml_equal                    9   different types not equal.
```

```
[OK]                parse_event_test_set          0   yaml/bomb.yml.
[OK]                parse_event_test_set          1   yaml/anchor.yml.
[OK]                parse_event_test_set          2   yaml/cohttp.yml.
[OK]                parse_event_test_set          3   yaml/linuxkit.yml.
[OK]                parse_event_test_set          4   yaml/seq.yml.
[OK]                parse_event_test_set          5   yaml/too_large.yml.
[OK]                parse_of_string               0   yaml/anchor.yml.
[OK]                parse_of_string               1   yaml/cohttp.yml.
[OK]                parse_of_string               2   yaml/linuxkit.yml.
[OK]                parse_of_string               3   yaml/seq.yml.
[OK]                parse_of_string               4   yaml/too_large.yml.
[OK]                reflect                       0   yaml/bomb.yml.
[OK]                reflect                       1   yaml/anchor.yml.
[OK]                reflect                       2   yaml/cohttp.yml.
[OK]                reflect                       3   yaml/linuxkit.yml.
[OK]                reflect                       4   yaml/seq.yml.
[ERROR]             reflect                       5   yaml/too_large.yml.
[OK]                emit                          0   emit.
[OK]                version                       0   version.
[OK]                quoted_scalars                0   quoted bool.
[OK]                quoted_scalars                1   quoted null.
[OK]                quoted_scalars                2   quoted float.
[OK]                quoted_scalars                3   quoted string.
[OK]                quoted_scalars                4   quoted int.
[OK]                quoted_scalars                5   plain int.
[OK]                quoted_scalars                6   plain bool.
[OK]                quoted_scalars                7   plain string.
[OK]                quoted_scalars                8   plain string with quote.
[OK]                quoted_scalars                9   plain string with int and quote.
[OK]                quoted_scalars               10   double quoted string.
[OK]                quoted_scalars               11   partially double quoted string.
[OK]                yaml_equal                    0   two null equal.
[OK]                yaml_equal                    1   two floats equal.
[OK]                yaml_equal                    2   two floats not equal.
[OK]                yaml_equal                    3   two bools equal.
[OK]                yaml_equal                    4   two strings equal.
[OK]                yaml_equal                    5   two strings not equal.
[OK]                yaml_equal                    6   two arrays equal.
[OK]                yaml_equal                    7   two arrays not equal.
[OK]                yaml_equal                    8   two objects equal.                                                                                                                                                                                                                    
[OK]                yaml_equal                    9   different types not equal. 
```

Third, (and most surprising) with some trivial changes to the logging, I was able to get the `parse_event_test` for `too_large.yml` to pass. This change was only to the pp_event code, which when switched to only print the position record, passed the test run. I have no idea why this would make any difference.

Fourth, in attempting to get the `reflect` test to work. I updated the `Stream.emitter` to accept the len. It was actually documented, but not implemented. However I wasn't able to get the test to work when manually increasing the len, just for the `too_large.yml` test.

### Next Steps
I'm currently unsure of any additional solutions, that don't involve major changes. So I'm still trying to shrink the problem. Hopefully, above approaches and results can help spark ideas from you.

--- a/ffi/bindings/yaml_bindings.ml
+++ b/ffi/bindings/yaml_bindings.ml
@@ -40,7 +40,7 @@ struct
     foreign "yaml_parser_delete" C.(ptr T.Parser.t @-> returning void)
 
   let parser_set_input_string =
-    foreign "yaml_parser_set_input_string" C.(ptr T.Parser.t @-> string @-> size_t @-> returning void)
+    foreign "yaml_parser_set_input_string" C.(ptr T.Parser.t @-> ptr char @-> size_t @-> returning void)
 
   let parser_parse =
     foreign "yaml_parser_parse" C.(ptr T.Parser.t @-> ptr T.Event.t @-> returning int)

--- a/lib/stream.ml
+++ b/lib/stream.ml
@@ -205,15 +205,17 @@ let get_version () =
 type parser = {
   p: T.Parser.t Ctypes.structure Ctypes.ptr;
   event: T.Event.t Ctypes.structure Ctypes.ptr;
-  buf: string;
+  buf: char Ctypes_static.carray;
 }
 
-let parser buf =
+let parser str =
   let p = Ctypes.(allocate_n T.Parser.t ~count:1) in
   let event = Ctypes.(allocate_n T.Event.t ~count:1) in
   let r = B.parser_init p in
-  let len = String.length buf |> Unsigned.Size_t.of_int in
-  B.parser_set_input_string p buf len;
+  let buf = Ctypes.CArray.of_string str in
+  let buf_ptr = Ctypes.CArray.start buf in
+  let len = String.length str |> Unsigned.Size_t.of_int in
+  B.parser_set_input_string p buf_ptr len;
   match r with
   | 1 -> R.ok {buf; p;event}
   | n -> R.error_msg ("error initialising parser: " ^ string_of_int n)

